bondcpp/src/bond.cpp

```diff
@@ -473,7 +473,9 @@ void Bond::onHeartbeatTimeout()
 {
   {
     std::unique_lock<std::mutex> lock(mutex_);
-    sm_.HeartbeatTimeout();
+    if (sm_.getState().getId() != SM::WaitingForSister.getId()) {
+      sm_.HeartbeatTimeout();
+    }
```

Comment:
@mjcarroll why the change? why the difference with `bondpy`?

Reply:
The `WaitingForSister` state doesn't have an implementation for `HeartbeatTimeout` as you can see here:  https://github.com/ros/bond_core/blob/5a79fecb3e8b6f2c210d83669b5cb40d47d62c47/bondcpp/include/bondcpp/BondSM_sm.h#L67-L79

This was causing a segfault in the `remoteNeverConnects` test: https://github.com/ros/bond_core/blob/fb0fa11895463dfa771a1aa329ff9113d59f4ee1/test_bond/test/test_callbacks_cpp.cpp#L98-L107

I don't believe that this impacts behavior, just guards against an undefined condition that wasn't previously exposed from the unit tests.

Reply:
I'm not super familiar with `bond`'s architecture, but if a heartbeat timeout can while waiting for sister, should said state handle that condition (even if ignoring it)? That applies to both `bondcpp` and `bondpy` (where I don't an `if` clause like this one).

Reply:
@mjcarroll bump (this is in the list of things to triage for Galactic)

Reply:
IIUC `ros2` `bondcpp` has this new method `start()` that [resets the heartbeat timer](https://github.com/ros/bond_core/blob/fb0fa11895463dfa771a1aa329ff9113d59f4ee1/bondcpp/src/bond.cpp#L377), while `kinetic-devel` `bondcpp` only [reset the heartbeat timer the first time someone called Heartbeat](https://github.com/ros/bond_core/blob/5a79fecb3e8b6f2c210d83669b5cb40d47d62c47/bondcpp/src/bond.cpp#L409-L412). It probably wasn't possible for the `WaitingForSister` state to see `HeartbeatTimeout` in ROS 1.

Maybe the heartbeat timer should be reset only when the connection is formed? I also agree with @hidmic that `WaitingForSister` should implement that callback too, but the implementation should be a warning or error that the callback is unexpected.